### PR TITLE
Enable CET shadow stack compatibility in `apphost`/`singlefilehost` by default

### PR DESCRIPTION
**Pull request by elinor-fung**
Enable CET shadow stack compatibility in apphost/singlefilehost by default
- Build Windows x64 `apphost` and `singlefilehost` with `/cetcompat` (shared libraries are already being built with this flag)
- Allow disabling CET compat (clearing the DLL characteristics bit in the PE image) in `HostWriter.CreateAppHost`
- Add unit/integration tests

This change makes it so that a user's app will have CET compatibility enabled by default (for Windows x64). There will be a corresponding change on the SDK side such that it can be disabled with `<CetCompat>false</CetCompat>`.

cc @dotnet/appmodel @janvorli @mangod9

### CODE REVIEW

File: src/installer/managed/Microsoft.NET.HostModel/AppHost/PEUtils.cs
Comment (mangod9) [inline src/installer/managed/Microsoft.NET.HostModel/AppHost/PEUtils.cs:57]: dont think there are multiple `IMAGE_DEBUG_TYPE_EX_DLLCHARACTERISTICS` sections right? should it break after unsetting the CET_COMPAT bit at the end?

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Review by mangod9 — APPROVED** [review]
Thanks for implementing this. We should keep an eye on CI, we had noticed that the CET specific CI was failing with new EH enabled, so wondering if that would affect a wider spectrum after this merges.

**Review by AaronRobinsonMSFT — APPROVED** [review]
:shipit:

File: src/native/corehost/apphost/standalone/CMakeLists.txt
Comment (jkoritzinsky) [inline src/native/corehost/apphost/standalone/CMakeLists.txt:51]: I'm not sure if this will add the flags to the target as the variables set by `add_linker_flag` only take effect on targets declared after the call (which is why the existing one is using such a weird syntax, it's adding it to the target after the fact)

Comment (jkoritzinsky) [reply src/native/corehost/apphost/standalone/CMakeLists.txt:51]: `target_link_options` is the more modern option, but if add_linker_flag works, then it's fine to use that.

**Comment by janvorli** [comment]
@elinor-fung please don't merge this change in yet. I was investigating a CET test leg failure yesterday and found a problem with the shadow stack and the new exception handling that needs to be addressed before this change gets in.

**Comment by janvorli** [comment]
@elinor-fung my CET fix is merged in, so please feel free to merge this in once the CI is green.

**Pull request merged**
